Ramping up gamification

kirubarajan opened the pull request:
This PR implements:
- Trophies on profile page
- Quality-of-life fixes for loading generations
- Ability to perform annotations without explicitly creating a new user (creates user with autogenerated username and no password that can both be set later)

Future work this PR has created:
- Need to refactor the navbar into it's own component to prompt user into setting password/creating an account
- Need to create functionality and figure out flow to have users set their new username/password
- Touch up trophies/come up with more trophy ideas

daphnei reviewed (changes requested):
Some regressions and other issues I noticed:

* When trying to complete annotations as an unlogged in user I am getting the following error (maybe just because I haven't updated my database?): 
```
Internal Server Error: /save/
Traceback (most recent call last):
  File "/mnt/castor/seas_home/d/daphnei/transformers/lib64/python3.6/site-packages/django/utils/datastructures.py", line 76, in __getitem__
    list_ = super().__getitem__(key)
KeyError: 'irrelevant'
```
* The navbar no longer displays the total number of annotations the user has one so far.
* Rather than having the two buttons "Play Now" and "Sign Up", the "Play Now" button should take user to a second page (you can re-use the existing login page) where they are asked if they want to make an account (mentioning the benefits), log in, or play without logging in. As it stands currently, the home page has no button saying "Log In" which is suboptimal.
* When I press "play now" and proceed to annotate without having made an account, the navbar should not present me with a button saying "Log out". That button should be replaced with one that says "Create account"
* The Profile page should be disabled for temporary users. The easy thing to do for now might be to just remove that button from the navbar if it is a temporary account.
* You should also check to make sure temporary users do not appear in the leaderboard.

kirubarajan commented on annotation/core/views.py:
```diff
@@ -99,45 +112,54 @@ def profile(request, username):
+        unseen_set = Generation.objects.all()
 
+        # TODO: allow temporary user to set password by checking user.has_usable_password()
+        user = User.objects.create(username=generate_random_username())
```
Yes, this is creating a new user which we keep track of by whether or not they have a password set.

daphnei replied:
I think it might make the code clearer to keep track of this by a boolean in the Profile object (even if that's redundant.)

daphnei replied:
Oh, I see you added this, nevermind!

kirubarajan commented on annotation/core/views.py:
```diff
@@ -147,17 +169,16 @@ def annotate(request):
         print("In annotate with qid = {}.".format(qid))
         generation = Generation.objects.get(pk=qid)
-        if seen_set.filter(generation=qid).exists():
+        if request.user.is_authenticated and seen_set.filter(generation=qid).exists():
```
True for everybody IIRC.

kirubarajan commented on annotation/populate_database.py:
```diff
@@ -146,15 +146,16 @@ def populate_db(generations_path, version):
-                playlist.generations.add(generation)
+                        gen_text=SEP.join(generation["generation"]))
+                    playlist.generations.add(generation)
+                except Exception as e:
```
The issue was a mismatch in encodings between model generations and the database. IIRC it was a negligible percentage of errors (< 2%).

Example error:
```
django.db.utils.OperationalError: (1366, "Incorrect string value: '\\xF0\\x9F\\x99\\x82_S...' for column 'body' at row 1")
```

kirubarajan commented on annotation/core/views.py:
```diff
@@ -35,6 +48,9 @@ def onboard(request):
 
 
 def splash(request):
+    if request.user.is_authenticated:
```
Resolved.

kirubarajan commented on annotation/core/views.py:
```diff
@@ -99,45 +112,54 @@ def profile(request, username):
-    available_set = counts.filter(count__gte=1,
-                                  count__lte=GOAL_NUM_ANNOTATIONS,
-                                  generation__in=unseen_set).values('generation')
+    counts = counts.filter(count__gte=1, count__lte=GOAL_NUM_ANNOTATIONS, generation__in=unseen_set).values('generation')
```
Renamed `counts` and `available_set` to `available_generations` and `generations`. The semantics of the `generations` variable name has good downstream consistencies (e.g. `generation = random.choice(generations)`).

daphnei replied:
Seems good!

kirubarajan commented:
Next PR: 
- adding page for converting temporary user to activated user
- refactoring into sub-apps for maintainability
- add status indicator for # of annotations completed

daphnei commented:
Hey Arun! Here are a few bugs I found:

* Pressing the "Play Now!" button from the home page goes straight to the "/annotate" endpoint without showing the Help instructions or allowing playlist selection. I would expect this button to take you to the "/onboard" endpoint.
* Pressing "Save Progress" takes you to the /play endpoint (the one that lets you switch playlists) rather than the endpoint to create an account. 
* There does not seem to be any way to log in anymore. The login button on the home page takes you to the "/play" endpoint.

kirubarajan commented:
>     * Pressing the "Play Now!" button from the home page goes straight to the "/annotate" endpoint without showing the Help instructions or allowing playlist selection. I would expect this button to take you to the "/onboard" endpoint.

Fixed by redirecting to the `onboard` route and performing temp user auth on that view.
 
>     * Pressing "Save Progress" takes you to the /play endpoint (the one that lets you switch playlists) rather than the endpoint to create an account.

This is going to be a different auth flow, and will be covered in the next PR. I've commented out the button until the next PR.

>     * There does not seem to be any way to log in anymore. The login button on the home page takes you to the "/play" endpoint

I made it clearer on the splash page what you should do to login/logout depending on your auth status.

>     * The auto-generated users show up on the leaderboard. For example, I see an entry with Username "i3ok-9scx-3jcq-pnda-4".

I've added a query to fix this, but the entire leaderboard computation should be moved to an offline process in a later PR.

daphnei commented:
I am going to merge this so that you're unblocked, but I found one more bug I want to note:

* When you have done some number of annotations as a not-logged in user, and then go back to the homepage, the button that should say "Log in" now says "log out"
